library/Freckle/App/Http/Cache.hs

```diff
@@ -0,0 +1,304 @@
+  , evict :: CacheKey -> m (Either SomeException ())
+  }
+
+data CachedResponse body = CachedResponse
+  { response :: Response body
+  , inserted :: UTCTime
+  , ttl :: CacheTTL
+  }
+  deriving stock (Show)
+
+isCachedResponseStale :: CachedResponse body -> UTCTime -> Bool
+isCachedResponseStale cached now =
+  addUTCTime (fromIntegral cached.ttl) cached.inserted < now
+
+-- Wrap a function from "Freckle.App.Http" with caching
+--
+-- Verify that the request is cacheable (e.g. a @GET@), then cache it at a
+-- derived key (from URL and considering any @Vary@ headers). The response will
+-- only be cached if @Cache-Control@ allows it. @Cache-Control@ is also used to
+-- determine TTL (e.g. @max-age@)
+--
+-- - <https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching#vary>
+-- - <https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching#fresh_and_stale_based_on_age>
+--
+-- If a cached response is stale, but it has an @ETag@ header, we will make the
+-- request using @If-None-Match@ and still return (and retain) that cached
+-- response if we receive a @304@ response.
+--
+-- - <https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag#caching_of_unchanged_resources>
+--
+-- Usage:
+--
+-- @
+-- import "Freckle.App.Http"
+-- import "Freckle.App.Http.Cache"
+-- import "Freckle.App.Http.Cache.Memcached"
+--
+-- data Foo = Foo
+--   { -- ...
+--   }
+--   deriving stock Generic
+--   deriving anyclass (Serialise, ToJSON)
+--
+-- getFooById
+--   :: ( 'MonadUnliftIO' m
+--      , 'MonadHttp' m
+--      , 'MonadLogger' m
+--      , 'MonadTracer' m
+--      , 'MonadReader' env m
+--      , 'HasMemcachedClient' env
+--      )
+--   => FooId -> m [Foo]
+-- getFooById fooId = do
+--   let req = parseRequest_ $ \"https:\/\/api.com\/foos\/\" <> fooId
+--   'httpCached' ('memcachedHttpCacheSettings' 'fiveMinuteTTL') 'httpJson' req
+-- @
+httpCached
+  :: forall m body t
+   . Monad m
+  => HttpCacheSettings m body t
+  -> (Request -> m (Response body))
+  -> Request
+  -> m (Response body)
+httpCached settings doHttp req =
+  maybe (doHttp req) handleCachableRequest $ getCachableRequestKey settings req
+ where
+  handleCachableRequest key = do
+    now <- settings.getCurrentTime
+    result <- fromEx Nothing $ settings.cache.get key
+
+    case result of
+      Nothing -> doHttpSet now key
+      Just val ->
+        case settings.codec.deserialise req val of
+          Left err -> doHttpSet now key <* settings.logWarn err
+          Right cresp | isCachedResponseStale cresp now -> do
+            case lookupHeader hETag cresp.response of
+              Nothing -> evictAndRecache now key =<< doHttp req
+              Just etag -> do
+                resp <- doHttp $ addRequestHeader hIfNoneMatch etag req
+                case statusCode (getResponseStatus resp) of
+                  304 -> pure cresp.response
+                  _ -> evictAndRecache now key resp
+          Right cresp -> pure cresp.response
```

Comment:
I tried writing this so many different ways, and none of them were particularly pleasing beyond this straight-forward stair-`case`-ing. I think there's really no hope of making this code super clear. Instead, I recommend reading the tests to understand the behaviors.